Fixed taking the correct interval and placing them into the correct position

Marc-Goritschnig opened the pull request:
The problem was with taking the max of the intervals, which were tuples of (x, y) values. We have to take the biggest starting interval, therefore biggest x, and then the smallest ending interval, meaning y. An example which did not work before and which put all comments in the beginning of the code after unparse is performed is:
`
if 1 == 1:
   #Test1
   print('a')
else:
   #Test2
   print('b')
print('b')
`
![image](https://github.com/t3rn0/ast-comments/assets/43956528/2455c1ae-a2a1-4298-9626-6fc4bfed40d2)

Marc-Goritschnig committed:
Fixed taking the correct interval

The problem was with taking the max of the intervals, which were tuples of (x, y) values. We have to take the biggest starting interval, therefore biggest x, and then the smallest ending interval, meaning y. An example which did not work before and which put all comments in the beginning of the code after unparse is performed is:

code = """
if 1 == 1:
   # Test1
   print('a')
else:
   # Test2
   print('b')
print('b')
"""

Commit:
Adding an extension section of ast nodes to populate the presence of the new comment nodes into the interval ranges of ast nodes.

Using this extension allows for comments to be correctly parsed in any position different than inlined comments

Added an additional test to parse and unparse tests, to cover all the comment placements possible

Commit:
updating also the end col offset corresponding to the new changed end line number

Commit:
a comment was still missed when we do not extend parent nodes intervals (before the min max of attributes was taken)

in the following case the last comment was not correctly parsed:

for i in [1,2,3]:
    for i2 in [1,2,3]:
        print('')
    # Comment
print('')

t3rn0 commented on ast_comments.py:
```diff
@@ -111,18 +113,69 @@ def _get_tree_intervals(
+def _post_extend_interval(
+        interval: Tuple[int, int, str],
+        code: str
+) -> Union[List[Tuple[int, int]], ast.AST]:
```
wrong return type

Marc-Goritschnig replied:
fixed (changed to tuple (int, int))

t3rn0 replied:
it's better to use `Tuple` instead of `tuple` if we want to support python3.8

t3rn0 commented on ast_comments.py:
```diff
@@ -111,18 +113,69 @@ def _get_tree_intervals(
 
+### Try to move lower bound lower and upper bound higher while not going out of bounds concerning the current block
+def _post_extend_interval(
+        interval: Tuple[int, int, str],
```
the last item of the tuple isn't used. we can remove it

Marc-Goritschnig replied:
i removed it and changed the calls accordingly

t3rn0 commented on test_parse.py:
```diff
@@ -8,6 +8,30 @@
 from ast_comments import Comment, parse
 
 
+def test_all_comment_places_in_tree():
```
Can you refactor the test, please? Current test does not test all-comments-in-place. It tests only that there are 7 nodes at the module level. 
When testing "Comment 4" position there's no need in Comment 1,2,7,8, and related prints.

I would split it to smaller tests covering specific details of your commit
Additionally, can you add test showing that if "Comment 4" has indent lower than "print('2')", then it goes to "else" block? This behavior is not a straightforward one so it better be documented somehow.

Marc-Goritschnig replied:
added tests for the several placements of comments which were not covered and removed the big test.

also added tests to show behaviour on wrong indentation of comments within tests

t3rn0 commented on ast_comments.py:
```diff
@@ -111,18 +113,69 @@ def _get_tree_intervals(
+        code: str
+) -> Union[List[Tuple[int, int]], ast.AST]:
+    lines = code.split('\n')
+    lines.insert(0, '')
```
please, add a comment about this "insert"

Marc-Goritschnig replied:
Added explanation this line is neccessary
 - it is easier to add a line than to use everywhere lineno - 1

t3rn0 commented on ast_comments.py:
```diff
@@ -111,18 +113,69 @@ def _get_tree_intervals(
 
 
+### Try to move lower bound lower and upper bound higher while not going out of bounds concerning the current block
+def _post_extend_interval(
```
Why "post"? There's no code with "main-part-extend-interval" functionality. We can call it simply "_extend_interval"

Marc-Goritschnig replied:
Changed the name

t3rn0 commented on ast_comments.py:
```diff
@@ -111,18 +113,69 @@ def _get_tree_intervals(
     return res
 
 
+### Try to move lower bound lower and upper bound higher while not going out of bounds concerning the current block
```
Add please to the comment that the extension is based on indentation patterns

Marc-Goritschnig replied:
Added additional information about the indentation usage

Commit:
Fixed:
- return type
- removed third item from tuple which was not used
- added comment to explain the insert(0, '') statement
- renamed post... function
- extended function comment for behaviour explanation

Other fixes come in the next commit

Commit:
Fixed behaviour when comments have wrong indentation at the beginning of an interval (block, e.g. if) by ignoring them when searching for the next lower bound

Commit:
added tests for the several placements of comments which were not covered and removed the big test.

also added a test to show behaviour on wrong indentation of comments within tests

Commit:
Fixed test not working due to off by one error in max line which is checked for extending the intervals

Marc-Goritschnig commented:
I reworked all your requested changes

t3rn0 commented on test_parse.py:
```diff
@@ -8,6 +8,60 @@
 
 
+def test_comment_at_start_of_inner_block_getting_correctly_parsed():
+    """Parsed tree has Comment node."""
```
Same docstring for all new tests

t3rn0 commented on test_unparse.py:
```diff
@@ -21,6 +21,57 @@ def _test_unparse(source: str):
 
+
+def test_comment_at_start_of_inner_block_getting_correctly_parsed():
+    """Parsed tree has Comment node."""
```
Same docstring for all new tests

t3rn0 reviewed (changes requested):
Please, run `make format && make lint` before next commit

Commit:
fixed doc and formatted files
and changed tuple to Tuple

Marc-Goritschnig commented:
i have changed those few things and formatted the file with intellijs pycharm, sadly i have not make tools currently and i cannot execute the script

if it is still necessary after the formatting i have done i would do that on the next weekend

t3rn0 commented:
Merged it. In the next commit I made a few minor formatting changes. Thank you for making this PR. I'll ship it to PyPi with version==1.1.1 this weekend
